### ISSUE TITLE
Automatic repair after alter table which changes the RF of a keyspace

### ISSUE THREAD

**Issue by asias** [scylladb/scylladb#12776]
When user increases the RF of a keyspace using alter table, e.g, from {dc1:2} to {dc1:3}, we need to run repair to sync the data to the new replica node. To automate the repair process needed by alter table, scylla core could monitor the alter table cmd and notify the management stack that a repair is needed and trigger the repair automatically.

**Comment by tzach** [comment]
Another alternative is moving the Keyspace and RF APIs from the DB level to a management element, which will initiate and monitor the RF change and Repair.

https://github.com/scylladb/scylladb/issues/8619

**Comment by dorlaor** [comment]
Tzach, it's better to keep it in-core and have the logic done in one place

**Comment by tzach** [comment]
> Tzach, it's better to keep it in-core and have the logic done in one place

Why?
Updating the RF is likely a result of adding/removing nodes,  so the entire process is already managed elsewhere, for example, by the Operator.

Once #8619 is done, updating the RF will be initiated by the admin API, not the data API.
(this does not invalidate either alternatives, but it demostrate there is another management element in place)

**Comment by bhalevy** [comment]
> @Benny Halevy ***@***.***> this is classical for the task manager)

The task manager is not intended to initiate anything but rather only to track and control tasks that other entities create.

In this case, I'd say that changing the replication strategy or replication factor affect the replication map - hence it is akin to a node operation and so it should be orchestrated using the topology changes infrastructure.

Keep in mind that reducing the replication factor requires cleanup, and that should be automated their too.

Cc @tgrabiec @kostja @kbr-scylla

**Comment by tgrabiec** [comment]
I agree with Benny that we should make RF changes go through a (safe) topology change machinery. That's the only way to guarantee that consistency invariants are maintained.

**Comment by asias** [comment]
Assume we have 10 nodes in the cluster.  No topology change is in progress.  If a user changes from

RF = 3, QUORUM READ/WRITE

(replica  set = n1, n2, n3. Only n1, n2 have the data)

to

RF = 5, QUORUM READ/WRITE

(replica set = n1, n2, n3, n4, n5. Only n1, n2 have the data)

quorum reads which goes to n3, n4, n5 gives wrong result immediately.

Without user changing the read consistency at the application level, scylla can do nothing to guarantee the consistency requirement all the time during the transition.

**Comment by kbr-scylla** [comment]
> quorum reads which goes to n3, n4, n5 gives wrong result immediately.

> Without user changing the read consistency at the application level, scylla can do nothing to guarantee the consistency requirement all the time during the transition.

This is Scylla's fault, not the user's.

Scylla should ensure - by streaming/repairing data as part of the ALTER command - that any successful quorum write will be observed by a later quorum read. So in your example, as part of the ALTER, it should ensure that at least one of n3/n4/n5 contains the data.

**Comment by tgrabiec** [comment]
> Without user changing the read consistency at the application level, scylla can do nothing to guarantee the consistency requirement all the time during the transition.

If this would go through our safe topology changes algorithm then consistency would be guaranteed. It works like this. Until the data is replicated, topology is transitional, and coordinators see that old replica set is {n1, n2, n3} and the new replica set is {n1, n2, n3, n4, n5}. When executing CL=QUORUM, they will require a quorum in both old and new replica set. For example, a read would be performed using a set which is a quorum in both, e.g. {n2, n3, n4}.
